PVR DMA implementation

andressbarajas opened the pull request:
### PVR and TA DMA Clarifications and New PVR DMA Functionality
It turns out that what we've been using for submitting textures and vertex data is TA DMA, not PVR DMA as previously thought. This PR adds true PVR DMA functionality to the pvr_dma_*() functions. Below is a breakdown of the capabilities of each DMA type and the changes introduced in this PR.

### DMA Types Overview
=== TA DMA ===
- Can DMA texture data.
- Can DMA vertex data.
- Direction: SH4 => TA only (one-way transfer).

=== PVR DMA ===

- Can DMA texture data.
- Can DMA PVR register data (can be used to DMA palettes and other register data).
- Direction: Bi-directional (SH4 <=> PVR), allowing both upload and download transfers.

### Breaking API Changes
EDIT:
This PR introduces no API-breaking changes but deprecates pvr_dma_ready() and provides 2 new functions to check ready status.

I moved the pvr_dma functions to their own header, pvr_dma.h.  The params to these functions are now more strongly typed with enums rather than ints.

andressbarajas committed:
Initial PVR DMA implementation. Blocking not working

andressbarajas committed:
Fix asic evt value for PVR DMA

DC-SWAT commented:
Have you thought about simply combining them into one API? The choice of DMA can be made based on the `type` of transfer (see `pvr_dest_addr()`). If you need to transfer to or from PVR RAM via the Root bus, then use DMA2, if you need to transfer to TA, then use TA DMA. Just needs some tweaking for`pvr_dma_transfer()` and that all.
And SQ's already can transfer to any memory, there is no point in copying their functions at all, they already do it by transfer `type`.

andressbarajas commented:
@DC-SWAT Yep! Thats the goal. I just have two implementations to test things out and have the ability to compare. Another function besides pvr_dma_transfer that we have to think about is pvr_dma_ready().  Not sure how to handle that. I created this PR to get input but mainly to share with @gyrovorbis since he is way forward thinking when it comes to APIs. Im expecting him to finish the work I started.

andressbarajas committed:
Merged pvr2_dma.c and pvr_dma.c

andressbarajas committed:
NA => DIR_NA

DC-SWAT commented on kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:
```diff
@@ -88,81 +145,137 @@ static uintptr_t pvr_dest_addr(uintptr_t dest, int type) {
+                            PVR_DMA_TO_SH4, callback, cbdata);
+}
+
+int pvr_dma_load_pal(void *sh4, pvr_ptr_t pvr, size_t count, int block,
```
Also can be applied to fog.

andressbarajas committed:
Fix ta dma check

andressbarajas committed:
Updated some documentation, spacing, and using a switch for type

andressbarajas committed:
Added tons of documentation.

andressbarajas committed:
More documentation updates

andressbarajas committed:
Revert API breakage

DC-SWAT commented on kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:
```diff
@@ -20,224 +20,395 @@
-    uintptr_t src_addr = ((uintptr_t)src);
+int pvr_dma_transfer(void *src, void *dest, size_t count, pvr_dma_mode_t type,
+                    int block, pvr_dma_callback_t callback, void *cbdata) {
+    uintptr_t src_addr = (uintptr_t)src;
```
```suggestion
    uintptr_t src_addr = ((uintptr_t)src) & MEM_AREA_CACHE_MASK;
```

DC-SWAT commented on kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:
```diff
@@ -20,224 +20,395 @@
+int pvr_dma_transfer(void *src, void *dest, size_t count, pvr_dma_mode_t type,
+                    int block, pvr_dma_callback_t callback, void *cbdata) {
+    uintptr_t src_addr = (uintptr_t)src;
+    uintptr_t dest_addr = (uintptr_t)dest;
```
```suggestion
    uintptr_t dest_addr = ((uintptr_t)dest) & MEM_AREA_CACHE_MASK;
```

DC-SWAT commented on kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:
```diff
@@ -20,224 +20,395 @@
+        }
+
+        /* Figure out which direction we are going */
+        if(((dest_addr & MEM_AREA_CACHE_MASK) >> 24) == 0x0c) {
```
```suggestion
        if((dest_addr >> 24) == 0x0c) {
```

andressbarajas committed:
Removed _pal and _fog dma functions for a possible future PR.  The code no longer breaks the API but deprecates pvr_dma_ready()

andressbarajas commented:
@DC-SWAT @gyrovorbis Its READY! Looking forward to full review.

DC-SWAT commented on kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:
```diff
@@ -20,224 +20,389 @@
+        /* Block if necessary */
+        if(block)
+            sem_wait(&pvr_dma_done);
+    } else {
```
```suggestion
    }
    else {
```

DC-SWAT replied:
BlueCrab wouldn't like this.

andressbarajas replied:
Done

DC-SWAT commented on kernel/arch/dreamcast/hardware/pvr/pvr_dma.c:
```diff
@@ -20,224 +20,389 @@
+        dbglog(DBG_ERROR, "%s: PVR DMA has not finished\n", func_name);
+        errno = EINPROGRESS;
+        return -1;
+    } else if(ta_dma->start != 0) {
```
```suggestion
    }
    else if(ta_dma->start != 0) {
```

andressbarajas replied:
Done

DC-SWAT reviewed (approved):
Looks good.

andressbarajas commented:
> Is there any way we might avoid having `pvr_dma_load_txr` and `pvr_txr_load_dma`? I get the reasoning, but it seems like it would be nothing but confusing. Maybe by shifting the naming to emphasize 'rb' (root bus) for PVR and 'ta' for the TA then deprecating `pvr_txr_load_dma` entirely?
> 
So `rb_dma_load_txr()` and `ta_dma_load_txr()`

andressbarajas committed:
Merge branch 'master' into pvr_dma_rootbus

andressbarajas committed:
Fix merge conflicts

andressbarajas committed:
Deprecate pvr_txr_load_dma(), use pvr_dma_ta_load_txr() instead

andressbarajas commented:
@QuzarDC @gyrovorbis @pcercuei It might be a while before #539 gets merged so I fixed the conflicts and renamed the confusing load texture functions. Ready for review.
